CMake refactor

jdumas opened the pull request:
Alright, here it is. As part of the efforts towards v3.0.0 described in #1696, this refactors libigl's CMake system to be more modern and maintainable. The new setup is inspired by [Lagrange](https://github.com/adobe/lagrange), e.g. how `include()` + `FetchContent` is used to download and handle third-party dependencies.

I wrote an upgrade guide [> here <](https://libigl.github.io/upgrade-v2-to-v3/), which I will update with final information once we merge this PR.

### Highlights
- CGAL and Boost are now built entirely from source via CMake. This means the CGAL/Cork module will work out of the box under Windows -- no need for a conda virtual environment. On Linux/macOS, the only dependency to install are GMP+MPFR.
- CMake targets and options have been renamed based on the module category: regular/copyleft/nonfree. This eliminates long module names (`igl::open_glfw_imgui`) that didn't bring much to the table, in favor of three root classes based on available licenses for each module.
- There is now a `LibiglOptions.cmake` sample file at the root of the project repository, which should make it easier to document/update available CMake options.
- There is now a unique entry point for the project, which is the root `CMakeLists.txt`. This facilitates maintenance, as before either `tutorial/CMakeLists.txt` and `cmake/libigl.cmake` could serve as potential entry points.
- The default behavior for module compilation is:
    - When compiling libigl on its own, all modules are enabled by default.
    - When compiling libigl as a subproject, all modules are disabled by default.
    - On Linux/macOS, if GMP/MPFR is not installed on the system, shows a warning and disable CGAL/Cork by default.
    - If the user explicitly enables those modules, but they do not have GMP/MPFR installed on the system, CMake configuration will still fail.
- Big cleanup of the tutorial CMake definitions. With the new setup adding a new tutorial is even easier!
- Cleaned up dependency setup for imgui/stb (among others)
- Unit tests have been split into different CMake target, based on which module is tested.
- Add support for Hunter + install() for the igl::core module (should be the same as before).
- Use a more generic `igl_copy_dll()` function to copy GMP/MPFR dlls on Windows.
- Cleaned up our .gitignore file.
- ~~Tested on macOS/Windows.~~ ~~Ok looks like I have some issues left to sort out on Windows, shouldn't be to bad.~~
  --> Looks like we're having compilations issues with Cork on Windows. Cork depends on gmpxx, which is a C++ wrapper around gmp types. The CGAL precompiled gmp libs only come with gmp itself. Thus supporting Cork with gmpxx on Windows would require integrating the mpir fork of gmp. Because this is a lot of work for little value, I vote that we **drop** support for the cork files altogether. After all we know that there are better alternatives nowadays.

### TODO
- Matlab/Mosek modules are currently missing, because I do not have them setup on my machine/have no way of testing them via Github Actions.
- `OSXFixDylibReferences.cmake` is also not used, for the same reason as above (although I am on macOS, I dot not have a setup that needs it).
- @rbsheth you wrote the initial hunterization of libigl. Could you let me know if you have any concern wrt to this new CMake setup.

alecjacobson commented:
How dependent are all these changes on each other? Each time I flip through this I get overwhelmed by the number of changes. I'm a bit worried that if we keep it as one larger PR it will never get merged. These changes may have more success if broken down into independent parts.

jdumas commented:
Well the goal of this PR was to refactor and cleanup our CMake build system. There are two parts to this 1) pulling the dependencies, and 2) cleaning up how modules are defined, as well as tutorials/unit tests. It may be a lot of changes, but it's really confined to the CMake scripts (I avoided doing changes to the .cpp unless I had to). I'd rather not do "half-cleanup" for this. Better look at it as a rewrite of the whole CMake build system where everything is new.

alecjacobson reviewed (commented):
Some initial comments. I think this will take a very long time to sift through everything.

jdumas commented on CMakeLists.txt:
```diff
@@ -1,74 +1,155 @@
+endif()
+
+# Check for gmp + mpfr and enable cork / cgal modules accordingly
+set(LIBIGL_DEFAULT_CGAL_CORK ${LIBIGL_TOPLEVEL_PROJECT})
```
We don't have an option `LIBIGL_DEFAULT_GMP_MPFR`. Here `LIBIGL_DEFAULT_CGAL_CORK` controls the default value of the options for the CGAL and Cork modules. If GMP and MPFR are not available on the target machine, then I disable the CGAL/Cork modules by default, since CGAL/Cork depend on GMP+MPFR.

alecjacobson replied:
Right, and I'm commenting that the name `LIBIGL_DEFAULT_CGAL_CORK` is confusing since it seems to primarily refer to the presence of gmp/mpfr.

jdumas replied:
The value of `LIBIGL_DEFAULT_CGAL_CORK` is set based on the presence of gmp/mpfr, but what it is used for is to set the default option for the CGAL and Cork modules. But hey I'm happy to rename it if you prefer.

jdumas replied:
One thing though: if libigl is not built as a subproject, `LIBIGL_DEFAULT_CGAL_CORK` will be OFF. Also on Windows, `LIBIGL_DEFAULT_CGAL_CORK` will always be ON by default for toplevel project. So if I rename the variable into `LIBIGL_GMP_MPFR_AVAILABLE`, this would be somewhat misleading (e.g. GMP/MPFR might be available, but because libigl is built as a subproject, it will be set to OFF).

alecjacobson replied:
I see. Maybe it makes sense to have a variable that tracks whether GMP_MPFR has been found and then the existing variables for wether building the cork / cgal modules.

jdumas replied:
> a variable that tracks whether GMP_MPFR has been found

What would you use it for? To track whether GMP/MPFR have been found, I would test for the existence of the imported targets `gmp::gmp` and `mpfr::mpfr`. Using variables in CMake is kinda brittle so I like to keep it to the bare minimum. But if you want to amend this part of the script to your liking feel free to give it a shot, since I'm not sure what you have in mind.

alecjacobson replied:
I don't think I can amend it because I don't understand it. What does `LIBIGL_DEFAULT_CGAL_CORK`  mean? `LIBIGL` I can guess means this is one of our variables, `DEFAULT` means this is a flag On or Off for a default setting. `CGAL_CORK` why both? Is this variable the same as `LIBIGL_DEFAULT_CGAL AND LIBIGL_DEFAULT_CORK`?

jdumas replied:
I mean it's just a variable name. The "correct" name should be:
```
LIBIGL_DEFAULT_VALUE_FOR_CMAKE_OPTION_CGAL_AND_CORK_MODULES_BASED_ON_WHETHER_GMP_AND_MPFR_ARE_AVAILABLE
```

Then I just use this variable twice to set the default value for the following CMake options:
- LIBIGL_COPYLEFT_WITH_CGAL
- LIBIGL_COPYLEFT_WITH_CORK

So really this should be invisible to the user. LIBIGL_DEFAULT_CGAL_CORK is just a "local variable", it's not a proper CMake option that the user has to concern themselves with. If the user provides their own value for LIBIGL_COPYLEFT_WITH_CGAL or LIBIGL_COPYLEFT_WITH_CORK, the user-provided option will take precedence.

jdumas commented on CMakeLists.txt:
```diff
@@ -1,74 +1,155 @@
-set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR})
-set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR})
-set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR})
+# Nonfree modules
```
Hmm so technically it seems that we have the following categories:
- Permissive: MIT-like license that allow commercial usage (igl::core, etc.)
- Copyleft: CGAL and other modules under GPL
- Noncommercial: Mosek and Matlab
- Proprietary: Triangle. Traditional copyright applies, and it is not clear how derivative work should be licensed (or even if we are allowed to produce derivative work by fixing bugs). Note that triangle has a proprietary license, even if the code itself is available online.

I think we should clearly indicate when a module is nonpermissive, such as copyleft or noncommercial/proprietary work. After all, all of libigl is available for academic usage, but I think it pays to be more transparent for non permissive works (like we do for copyleft).

Restricted might be a good alternative to encompass both noncommercial/proprietary licenses.

alecjacobson replied:
I agree the names should indicate these things. I'm happy with restricted_use unless we can thing of something better.

I wish we had a better way to indicate with these that it's the _dependency_  that restricts the use, not the libigl code. For example, with matlab, it should be hypothetically possible to link to octave (free alternative; I've never tried this and have no clue if it works).

jdumas replied:
I'd prefer `restricted` rather than `restricted_use` though, as it's just one word and less verbose to type/setup. Would that be acceptable?

For Octave bindings it would be interesting to support that, as we could add that to our Github Actions integration and make sure it works for everyone. (But I have very little interest in doing that myself though).

alecjacobson replied:
I'm OK with `restricted` as long as we clearly spell out in the cmake comments and the documentation what is restricted and how (basically your categories above).

jdumas replied:
Sure, I'll update that shortly.

jdumas commented on LibiglOptions.cmake.sample:
```diff
@@ -0,0 +1,72 @@
+# In order to persistently set default options for your project, copy this file
```
This file is just a convenience to set default options when building libigl as a standalone. When including libigl as a subproject, just set options like before. But now rather than passing a slew of options via command-line when building libigl, you can use this non-versioned file as an alternative.

alecjacobson replied:
I'm not sure I get the comparison to the current situation. Should I be looking at https://github.com/libigl/libigl-example-project/blob/next/CMakeLists.txt for an example of use? That looks pretty clean and doesn't seem to require copying this .sample file.

jdumas replied:
There are two use-cases:
1. You are cloning https://github.com/libigl/libigl.git and want to compile/try out tutorials.
2. You are writing a project using libigl, following the model of https://github.com/libigl/libigl-example-project

For the first case, this sample option file offers a way to customize what modules are compiled, and whether we are compiling in static mode or header-only. This is mostly a convenience for libigl developers, but can be useful to end-users as well. Previously, the only way to set options for this use-case is to pass them via command-line, or set them via CMake-GUI/ccmake.

For the second case, there is no need to copy the sample file, as libigl options are controlled by the parent project ([this line](https://github.com/libigl/libigl-example-project/blob/23263bde8786f7454f90709f55aca460376ba9b6/CMakeLists.txt#L7) as you have pointed out).

alecjacobson replied:
Ok, thanks, now I get it. Perhaps we can include this distinction in the comments/documentation if it's not already there.

jdumas commented on cmake/find/FindGMP.cmake:
```diff
@@ -0,0 +1,76 @@
+# Try to find the GNU Multiple Precision Arithmetic Library (GMP)
```
GMP/MPFR are a pain yes. We are not building them from source since they do not use a CMake build system. So the only option is to install via a package manager on macOS/Linux, or download a pre-built dll on Windows (which this find_package() will find).

jdumas commented on cmake/igl/modules/copyleft/cgal.cmake:
```diff
@@ -0,0 +1,32 @@
+# 1. Define module
```
I am wary of making _everything_ into a function. This file still separates the different parts of defining a libigl module based on:
1. Module name
2. Include headers
3. Target sources
4. Target dependencies
5. Unit test definitions

Imho abstracting away some of these steps would make it too much look like "cmake magic" (e.g. where does the include header definition come from?).

jdumas commented on include/igl/opengl/glfw/imgui/ImGuiMenu.cpp:
```diff
@@ -185,7 +185,7 @@ IGL_INLINE void ImGuiMenu::draw_viewer_menu()
   if (ImGui::CollapsingHeader("Workspace", ImGuiTreeNodeFlags_DefaultOpen))
   {
-    float w = ImGui::GetContentRegionAvailWidth();
+    float w = ImGui::GetContentRegionAvail().x;
```
I have added the `IMGUI_DISABLE_OBSOLETE_FUNCTIONS` define macro when updating our imgui dependency, which required updating the code accordingly. I could make a separate PR for this but I don't think it's worth the effort (adding `IMGUI_DISABLE_OBSOLETE_FUNCTIONS` to our current CMake would create conflict, and otherwise the change is not needed).

alecjacobson commented:
How does this cmake refactor affect the ability to link projects to prebuilt binaries (https://github.com/libigl/libigl/issues/1676)? I'm personally using this exclusively now to cut down on unnecessary builds. IIRC there was an idea that if we refactored the cmake (https://github.com/libigl/libigl/issues/1696) it would make package management systems easier. Is that what Hunter is? Or is this simply handled by the `install` option?

jdumas commented:
> How does this cmake refactor affect the ability to link projects to prebuilt binaries (#1676)? I'm personally using this exclusively now to cut down on unnecessary builds. IIRC there was an idea that if we refactored the cmake (#1696) it would make package management systems easier. Is that what Hunter is? Or is this simply handled by the install option?

This refactor should not affect your ability to link pre-built binaries of libigl. C++ files are mostly untouched, and I've only changed a few libs, e.g. there is now a `libigl_copyleft_core.a` rather than bundling those functions into the main `libigl.a`.

The next steps on my refactoring todo-list are as follows:
1. Move the .h/.cpp files into one subfolder per module, as discussed in #1696
2. Investigate packaging solutions using the following package manager:
     - Hunter
     - Conan
     - Vcpkg

For package manager we will start by packaging only the main igl_core module, but then once we have a good process in place we can look into extending support to other modules. As you can see in this PR there is a subfolder `cmake/recipes/hunter` that is paving the way to support Hunter packaging for other libigl modules as well. It's gonna take some time but I'll get there eventually.

alecjacobson commented:
We're a bit inconsistent about where `core` shows up in the names of things (library names, cmake module names, other places(?)). Do we _need_ `core` anywhere? For example we write `igl::core`. Can this be simply `igl`?

jdumas commented:
> We're a bit inconsistent about where core shows up in the names of things (library names, cmake module names, other places(?)). Do we need core anywhere? For example we write igl::core. Can this be simply igl?

Well, if you think of libigl in terms of modules, "core" is just the name of the "main" modules, i.e. where everything is directly in the namespace `igl::`, and in `#include <igl/xxx.h>`. In terms of CMake, it really simplifies the code to have everything follow the `igl_xxx` naming convention. Also, there are good reasons to use CMake target names with a semicolon in them (`igl::xxx` rather than `igl`) as it ensure the name corresponds to a proper CMake target and not just the name of a file on disk (i.e. it helps finding errors in your CMake code). Think of `igl::` as the namespace in terms of CMake targets, so the "core" libigl module would be named as `igl::core` (another possibility is `igl::igl`, but I find it redundant and less informative, as everything is already igl something).

Then there is the question of how to name the functions living directly under `igl/copyleft/`, such as `quadprog` and `progressive_hulls`. By extension, I felt that `igl_copyleft::core` was a natural naming convention for this module.

jdumas commented:
Alright I have renamed nonfree -> restricted + added details in comments, and updated the description in the ".sample" CMake file. Once this is merged I will update instructions on the website  accordingly.

rbsheth commented:
I haven't got a chance to review the PR in its entirety but happy to help test as needed.

jdumas commented:
> I haven't got a chance to review the PR in its entirety but happy to help test as needed.

Thanks! I just want to make sure this is not breaking existing Hunter integration. I don't have much experience packaging libraries with Hunter, and I'm not sure if you use the upstream repo or a clone for the Hunter package. If you have any insight please let me know.

jdumas commented on CMakeLists.txt:
```diff
@@ -1,74 +1,74 @@
-if(LIBIGL_TOPLEVEL_PROJECT)
-	# Set folders for Visual Studio/Xcode
-	igl_set_folders()
+    # include(CTest)
```
Why is this commented?

melMass commented:
Hey, I'm interested in adding Conan support (#1634) and it seems this PR is almost done and awaits review?

Thanks

jdumas commented:
Yes I'm waiting for green light from @alecjacobson and @danielepanozzo, but not sure when it's gonna happen...

melMass commented:
Thanks I'll wait for  @alecjacobson and @danielepanozzo to chime in!

alecjacobson commented:
🎉
